### ISSUE TITLE
'empty' encoding of enum types

### ISSUE THREAD

**Issue by monetdb-team** [MonetDB/MonetDB#825]
Date: 2006-01-10 01:55:54 +0100
From: thijs &lt;<thijswes>&gt;
To:   Stefan Manegold &lt;<Stefan.Manegold>&gt;
Version: -- development

Last updated: 2006-01-11 02:49:08 +0100

## Comment 4777
Date: 2006-01-10 13:55:54 +0100
From: thijs &lt;<thijswes>&gt;

When using enum types to encode strings, some of them
 seem to get the same internal representation that is
 used for 'not found'. This means that encoding this
 particular strings gives the same results as encoding
 strings that are not present in the enum table.
 Obviously, this may lead to unexpected behaviour.
 
 I tried to compose a minimal example, but did not
 succeed in making one where the error occurs. It *does*
 occur in the following example based on the
 [void,str]-bat in the attached file.
 
 
 module(enum);
 
 var entities := new(void,str).import("entities.txt");
 
 var enum_entities:=enum_create("enum_entities",entities);
 
 encode(enum_entities,"p").print();
  This produces, the expected: [ p ]
 
 encode(enum_entities,"p0").print();
  This produces, an empty representation,
  as expected (p0 is not in the table): [  ]
 
 encode(enum_entities,"p1").print();
  But, this also produces an empty representation,
  while "p1" is in the table: [  ]

## Comment 4778
Date: 2006-01-10 14:08:24 +0100
From: Niels Nes &lt;<niels>&gt;

Logged In: YES 
 user_id=43556
 
 The file with the bat is missing, could you give it another try?

## Comment 4779
Date: 2006-01-11 02:34:47 +0100
From: Stefan Manegold &lt;<Stefan.Manegold>&gt;

Logged In: YES 
 user_id=572415
 
 test script added as
 MonetDB/tests/BugTracker/Tests/empty_enum.SF-1401528.*
 

## Comment 4780
Date: 2006-01-11 02:41:13 +0100
From: Stefan Manegold &lt;<Stefan.Manegold>&gt;

Logged In: YES 
 user_id=572415
 
 the (extended) test script reveals that encoding fails only
 for this one value ("p1"), which happens to by the the 129th
 values that is encoded (i.e., assigning IDs starting with 0,
 this one get's 128(!)) --- maybe this helps for further
 debugging...
 

## Comment 4781
Date: 2006-01-11 14:49:07 +0100
From: Stefan Manegold &lt;<Stefan.Manegold>&gt;

Logged In: YES 
 user_id=572415
 
 fixed priciple design flaw of the enum encoding in CVS:
 
 (Thanks to Niels for the crucial hit!)
 
 Using internal enum IDs 0..255 (unsigned char) and 0..65535
 (unsigned short)
 bears the "risk" of hitting 128 (unsigned char) == -128
 (signed char) == chr_nil
 or 32768 (unsigned short) == -32768 (signed short) ==
 shr_nil; which eventually
 breaks the enum encoding (especially, since IDs chr_nil and
 sht_nil are explicitely
 used for non-encoded/non-exiting values...).
 
 To solve the problems, we shift the internal enum IDs to
 -127..127 (signed char)
 and -32767..32767 (signed short), savely excluding NIL from
 the (valid) ID ranges.
 
 NOTE:
 
 Since MonetDB implements NIL as the lowers value with the
 domain of its (signed)
 data types, using the full ranges of the same type as
 unsigned values locates
 NIL in the middle of the unsigned ranges, which is bound to
 trigger errors unless
 special care is taken.
 In other words, unsigned data types in MonetDB should (can)
 only be used with half
 the expected range (cf. OID)!
 

## Comment 4782
Date: 2010-05-04 09:32:09 +0200
From: Pseudo user for Sourceforge import &lt;<sfimport>&gt;

This bug was previously known as tracker item 1401528 at http://sourceforge.net/support/tracker.php?aid=1401528
